lib/common/zstdmt_priority_threadpool.h

```diff
@@ -0,0 +1,59 @@
+#ifndef ZSTDMT_PRIORITY_THREADPOOL_H_MODULE
+#define ZSTDMT_PRIORITY_THREADPOOL_H_MODULE
+
+#include <stddef.h>
+
+/* Priority thread pool context */
+typedef struct ZSTDMT_pThreadPoolCtx_s ZSTDMT_pThreadPoolCtx_t;
+
+/* Callback function for each job */
+typedef void (*ZSTDMT_pThreadPool_fn)(void* data);
+
+/* Basic Usage:
+ * ctx = ZSTDMT_pThreadPoolCtx_create(QUEUE_SIZE, NB_THREADS);
+ * for (i = 0; i < nbJobs; ++i)
+ *    ZSTDMT_pThreadPool_addJob(ctx, priority(i), 0, callbackFn, NULL);
+ * ZSTDMT_pThreadPool_waitAllThreads(ctx);
+ * ZSTDMT_pThreadPool_free(ctx); */
+/***************************************
+* ZSTDMT_pThreadPool API
+***************************************/
+
+/* Create a priority thread pool of 'queueSize' using 'nbThreads' worker threads
+ * @return : ptr to ZSTDMT_pThreadPoolCtx_t for thread pool */
+static ZSTDMT_pThreadPoolCtx_t* ZSTDMT_pThreadPoolCtx_create(size_t queueSize,
+  size_t nbThreads);
+
+/* Destroies the context */
+static size_t ZSTDMT_pThradPoolCtx_free(ZSTDMT_pThreadPoolCtx_t* ctx);
+
+/* Enqueue a job into the priority queue of the thread pool with given 'priority'
+ * Note: time complexity will be log(#jobs)
+ * Note: lower priorities are going to be excuted first (ie. priority 0 jobs will come first)
```

Comment:
Is there a maximum priority value ?

Reply:
Also : what happens when 2 jobs have the same priority value ?
Is there a specific order guaranteed ?

Reply:
Yes, whatever queueSize is supplied in create(). If every job is a new priority, that will be the max. Or do you mean absolute maximum priority value?

Reply:
I wasn't planning on maintaining order when priority values clash. That should be just fine for the way I'm planning on using this in the multi-threaded decoder. Jobs that have the same priority value will not require a particular execution order.